educator/src/Dscp/DB/SQLite/Types.hs

```diff
@@ -1,56 +1,61 @@
 module Dscp.DB.SQLite.Types
        ( -- * SQLite bindings
-         SQLiteRealParams (..)
-       , srpPathL
-       , srpConnNumL
-       , srpMaxPendingL
-       , SQLiteDBMode (..)
-       , _SQLiteReal
-       , _SQLiteInMemory
+         SQLiteRealParams
+       , SQLiteRealParamsRec
+       , SQLiteRealParamsRecP
+
+       , SQLiteDBMode
+       , SQLiteDBModeRec
+       , SQLiteDBModeRecP
+
+       , SQLiteParams
+       , SQLiteParamsRec
+       , SQLiteParamsRecP
+
        , SQLiteDB (..)
-       , SQLiteParams (..)
-       , sdpModeL
        ) where
 
 import Control.Concurrent.Chan (Chan)
-import Control.Lens (makeLensesWith, makePrisms)
-import Data.Aeson (FromJSON (..))
-import Data.Aeson.Options (defaultOptions)
-import Data.Aeson.TH (deriveFromJSON)
 import Database.SQLite.Simple (Connection)
-
-import Dscp.Util
+import Loot.Config ((::+), (:::), (::-), Config, PartialConfig)
 
 ----------------------------------------------------------
 -- SQLite bindings
 ----------------------------------------------------------
 
-data SQLiteRealParams = SQLiteRealParams
-    { srpPath       :: !FilePath
+type SQLiteRealParams =
+   '[ "path"       ::: FilePath
       -- ^ Path to the file with database.
-    , srpConnNum    :: !(Maybe Int)
+    , "connNum"    ::: Maybe Int
       -- ^ Connections pool size.
-    , srpMaxPending :: !Int
+    , "maxPending" ::: Int
       -- ^ Maximal number of requests waiting for a free connection.
-    } deriving (Show, Eq)
+    ]
+
+type SQLiteRealParamsRec = Config SQLiteRealParams
+type SQLiteRealParamsRecP = PartialConfig SQLiteRealParams
 
-makeLensesWith postfixLFields ''SQLiteRealParams
 
 -- | Database mode.
-data SQLiteDBMode
-    = SQLiteReal !SQLiteRealParams
+type SQLiteDBMode =
+   '[ "real"     ::- SQLiteRealParams
       -- ^ In given file using given number of connections
-    | SQLiteInMemory
+    , "inMemory" ::- '[]
       -- ^ In memory
-    deriving (Show, Eq)
+    ]
+
+type SQLiteDBModeRec = Config SQLiteDBMode
+type SQLiteDBModeRecP = PartialConfig SQLiteDBMode
 
-makePrisms ''SQLiteDBMode
 
-data SQLiteParams = SQLiteParams
-    { sdpMode :: SQLiteDBMode
-    } deriving (Show, Eq)
+-- | Wrapper for 'SQLiteDBMode', this helps passing the vinyl record around
+type SQLiteParams =
```

Comment:
One more thing I was just thinking: in most cases there is not need for both the "wrapper" and the "content only", so it may be a good idea to just define the second one inside the first.
At least we would avoid making unnecessary similarly-named types.

Reply:
Okay, got it, the problem was simple. I was simply confused by the fact you used type synonyms for lists representing branches, because it felt natural to me to just inline them, and I thought that you for some reason _had_ to use them, which made no sense. Now you've just inlined the inner lists with branches and everything became clear.